envoy/common/optref.h

```diff
@@ -13,76 +13,125 @@ namespace Envoy {
 
   /**
    * Helper to convert a OptRef into a pointer. If the optional is not set, returns a nullptr.
    */
-  T* ptr() {
-    if (this->has_value()) {
-      T& ref = **this;
-      return &ref;
-    }
-
-    return nullptr;
-  }
+  T* ptr() { return ptr_; }
 
   /**
    * Helper to convert a OptRef into a pointer. If the optional is not set, returns a nullptr.
    */
-  const T* ptr() const {
-    if (this->has_value()) {
-      const T& ref = **this;
-      return &ref;
-    }
+  const T* ptr() const { return ptr_; }
+
+  /**
+   * Helper to convert a OptRef into a ref. Behavior if !has_value() is undefined.
+   */
+  T& ref() { return *ptr_; }
+  const T& ref() const { return *ptr_; }
+
+  /**
+   * Helper to dereference an OptRef. Behavior if !has_value() is undefined.
+   */
+  T& operator*() { return *ptr_; }
+  const T& operator*() const { return *ptr_; }
+
+  /**
+   * @return true if the object has a value.
+   */
+  bool has_value() const { return ptr_ != nullptr; }
 
-    return nullptr;
+  /**
+   * @return true if the object has a value.
+   */
+  bool operator!() const { return ptr_ == nullptr; }
+  operator bool() const { return ptr_ != nullptr; }
+
+  /**
+   * Copies the OptRef into an optional<T>. To use this, T must supply
+   * an assignment operator.
+   *
+   * It is OK to copy an unset object -- it will result in an optional where
+   * has_value() is false.
+   *
+   * @return an optional copy of the referenced object (or nullopt).
+   */
+  absl::optional<T> copy() const {
+    absl::optional<T> ret;
+    if (has_value()) {
+      ret = *ptr_;
+    }
+    return ret;
   }
 
-  T& ref() { return **this; }
+  /**
+   * Places a reference to the an object into the OptRef.
+   *
+   * @param ref the object to be referenced.
+   */
+  void emplace(T& ref) { ptr_ = &ref; }
 
-  const T& ref() const { return **this; }
+  /**
+   * The value method has no intrinsic value to OptRef, but is left here for
+   * compatibility reasons. This is used in call-sites which would be needed if
+   * they were using optional<reference_wrapper<T>> directly without
+   * OptRef. Having this function makes it easier upgrade to using OptRef
+   * without having to change all call-sites.
+   *
+   * Usage of this method in the Envoy codebase appears to expect
+   * const-stripping: We are expecting to get a non-const T& out of a const
+   * OptRef<T>&. This should be investigated. It was not necessary when OptRef
+   * inherited from absl::optional<std::reference_wrapper<T>>. One place
+   * where this const-stripping expectation occurs is
+   * source/common/http/async_client_impl.cc where the constructor has an
+   * arg const AsyncClient::StreamOptions& options, that is used to populate
+   * a non-const Config reference.
+   *
+   * This must be called with has_value() true.
+   *
+   * @return a reference_wrapper around the value.
+   */
+  std::reference_wrapper<T> value() const { return std::reference_wrapper<T>(*ptr_); }
```

Comment:
I think constness is not right here. Probably needs some tests. I would expect `std::reference_wrapper<const T> value() const`

Reply:
Agreed. See the comment. I think the usage is wrong and I'm not sure how that compiles with the existing impl based on `optional<reference_wraper>`

Reply:
I think this is addressed now. Resolving.